File: 백준/Silver/2615. 오목/오목.py
"""
왼쪽에서 오른쪽, 위에서 아래쪽 방향으로 확인
↙, ↘, →, ↓
"""
board = [list(map(int, input().split(' '))) for _ in range(19)]

def is_in(y, x):
    return 0 <= y < 19 and 0 <= x < 19

def check_stone(y, x, color, mul_y, mul_x):
    global board
    ret = 0
    for i in range(19):
        ny, nx = y + mul_y * i, x + mul_x * i
        if not is_in(ny, nx) or (is_in(ny, nx) and color != board[ny][nx]):
            break
        ret += 1
        
    for i in range(1, 19):
        ny, nx = y - mul_y * i, x - mul_x * i
        if not is_in(ny, nx) or (is_in(ny, nx) and color != board[ny][nx]):
            break
        ret += 1
        
    return ret == 5

def solution():
    global board
    dy, dx = [1, 1, 0, 1], [-1, 1, 1, 0]
    for y in range(19):
        for x in range(19):
            color = board[y][x]
            for i in range(4):
                if color and check_stone(y, x, board[y][x], dy[i], dx[i]):
                    print(color)
                    if i == 0:
                        print(y + 1 + dy[i] * 4, x + 1 + dx[i] * 4)
                    else:
                        print(y + 1, x + 1)
                    return
    print(0)
                
solution()

File: 백준/Silver/2615. 오목/오목.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main{
    private static int[][] board;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        board = new int[19][19];

        for(int i=0 ; i<19 ; i++){
            StringTokenizer stk = new StringTokenizer(br.readLine());
            for(int j=0 ; j<19 ; j++){
                board[i][j] = Integer.parseInt(stk.nextToken());
            }
        }

        solution();
    }

    private static void solution(){
        for(int i = 0 ; i<19 ; i++){
            for(int j=0 ; j<19 ; j++){
                if(board[i][j] == 0)
                    continue;

                int exist = existWinner(i, j);
                if(exist == 0)continue;

                System.out.println(board[i][j]);
                if(exist == 1){
                    System.out.println((i + 1) + " " + (j + 1));
                }
                else if(exist == -1){
                    System.out.println((i + 5) + " " + (j - 3));
                }
                return;
            }
        }

        System.out.println(0);
    }

    private static int existWinner(int y, int x){

        if(check(y, x, 1, -1) + check(y, x, -1, 1)== 4)return -1;
        if(check(y, x, 1, 1) + check(y, x, -1, -1) == 4)return 1;
        if(check(y, x, 0, 1) + check(y, x, 0, -1) == 4)return 1;
        if(check(y, x, 1, 0) + check(y, x, -1, 0)== 4)return 1;
        
        return 0;
    }

    private static boolean isIn(int y, int x){
        return 0 <= y && y < 19 && 0 <= x && x < 19;
    }

    private static int check(int y, int x, int mulY, int mulX){
        int ret = 0, original = board[y][x];

        for(int i=1 ; i<20 ; i++){
            int ny = y + mulY * i, nx = x + mulX * i;
            if(!isIn(ny, nx) || board[ny][nx] != original)break;
            ret++;
        }

        return ret;
    }
}